python/ray/data/_internal/compute.py

```diff
@@ -258,11 +258,42 @@ def _apply(
         # Bin blocks by target block size.
         if target_block_size is not None:
             _check_batch_size(blocks_in, target_block_size, name)
-            block_bundles = _bundle_blocks_up_to_size(
-                blocks_in, target_block_size, name
-            )
+            block_bundles: List[
+                Tuple[Tuple[ObjectRef[Block]], Tuple[BlockMetadata]]
+            ] = _bundle_blocks_up_to_size(blocks_in, target_block_size, name)
         else:
-            block_bundles = [((b,), (m,)) for b, m in blocks_in]
+            block_bundles: List[
+                Tuple[Tuple[ObjectRef[Block]], Tuple[BlockMetadata]]
+            ] = [((b,), (m,)) for b, m in blocks_in]
+
+        # If the max number of the actor pool is set, the number of bundles should
+        # always be less than this max_size.
+        # Otherwise, it leads to inefficiencies with creating extra actor tasks and
+        # prevents the actor task from doing optimizations
+        # such as batch or block prefetching.
```

Comment:
A more orthogonal, future looking thing: a target bundle size that might be better than the user-provided `batch_size` is probably something like the following formula:
```python
target_size = max(
    min((prefetch_batches + 1) * batch_size_in_bytes, ctx.target_min_block_size),
    ctx.target_max_block_size,
)
```
I.e. where we bundle up to **at least** the `ctx.target_min_block_size` (default is 1 MiB) since that's what we consider to be the smallest "reasonable" block to make the task overhead worth it; if the number of desired concurrent batches is larger than this (e.g. batch size is larger than this and/or aggressive prefetching is specified), then we use that as a bundling target. And all of this is capped by `ctx.target_max_block_size`.

We'd probably still have the max actor pool size serve as a cap on the number of block bundles as well, but I'd imagine that the initial actor pool size (i.e. actors started at the beginning of execution) and the scale-up rate would be influenced by the number of block bundles.

We should experiment with a few of these hints/policies in the new execution model.